core/src/main/java/org/mule/processor/chain/ProcessorIterator.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.processor.chain;
+
+import org.mule.MessageExchangePattern;
+import org.mule.OptimizedRequestContext;
+import org.mule.VoidMuleEvent;
+import org.mule.api.MessagingException;
+import org.mule.api.MuleEvent;
+import org.mule.api.component.Component;
+import org.mule.api.endpoint.OutboundEndpoint;
+import org.mule.api.processor.MessageProcessor;
+import org.mule.api.service.Service;
+import org.mule.api.transformer.Transformer;
+import org.mule.api.transport.NonBlockingResponseReplyToHandler;
+import org.mule.execution.MessageProcessorExecutionTemplate;
+import org.mule.processor.AbstractInterceptingMessageProcessorBase;
+import org.mule.routing.MessageFilter;
+
+import java.util.List;
+
+/**
+ * Iterates over a list of {@link org.mule.api.processor.MessageProcessor}'s processing them one by one using the result
+ * of the first processor to invoke the second and so on.  MessageProcessor implementations aside from simply iterating
+ * over processors implement rules regarding if and when iteration should stop early or even stop temporarily and be
+ * continued later.
+ * <p/>
+ * This implementation processes each {@link org.mule.api.processor.MessageProcessor} in sucession in the same thread
+ * until or processors have been invoked or one of the following is returned by a processor:
+ * <li>{@link org.mule.VoidMuleEvent}</li>
+ * <li><code>null</code></li>
+ */
+public class ProcessorIterator
```

Comment:
Name is misleading to me, I would expect a ProcessIterator to implement Iterator, and that is not the case